src/combinatorics/additive/salem_spencer.lean

```diff
@@ -0,0 +1,386 @@
+is_O_iff_is_O_with.2 ⟨1, trivial_roth_bound'⟩
+
+end roth_number_nat
+
+/-!
+### Explicit values
+
+Some lemmas and calculations of the Roth number for (very) small naturals.
+-/
+
+section explicit_values
+
+/-- A simpler `decidable` instance for Salem-Spencer sets of naturals. We use it to prove small
+values of the Roth number by `rfl`/`dec_trivial`. -/
+def add_salem_spencer.decidable_nat {s : finset ℕ} : decidable (add_salem_spencer (s : set ℕ)) :=
+decidable_of_iff (∀ (a ∈ s) (b ∈ s), a < b → b + (b - a) ∉ s) begin
+  rw add_salem_spencer_iff_eq_right,
+  refine ⟨λ hs a b c ha hb hc habc, _, λ hs a ha b hb hab h, _⟩,
+  { by_contra h,
+    obtain hac | hac := lt_or_gt_of_ne h,
+    { refine hs _ ha _ hc hac _,
+      rwa [←add_tsub_assoc_of_le hac.le, ←habc, add_tsub_cancel_left] },
+    { have hbc : b < c := lt_of_not_ge' (λ h, (add_lt_add_of_lt_of_le hac h).ne' habc),
+      refine hs _ hb _ hc hbc _,
+      rwa [←add_tsub_assoc_of_le hbc.le, ←habc, add_tsub_cancel_right] } },
+  { refine hab.ne (hs ha h hb _),
+    rw [←add_tsub_assoc_of_le hab.le, add_tsub_cancel_of_le (le_add_left hab.le)] }
+end
+
+local attribute [instance] add_salem_spencer.decidable_nat
+
+@[simp] lemma roth_number_nat_zero : roth_number_nat 0 = 0 := rfl
+@[simp] lemma roth_number_nat_one : roth_number_nat 1 = 1 := rfl
+@[simp] lemma roth_number_nat_two : roth_number_nat 2 = 2 := rfl
+@[simp] lemma roth_number_nat_three : roth_number_nat 3 = 2 := rfl
+@[simp] lemma roth_number_nat_four : roth_number_nat 4 = 3 := rfl
+@[simp] lemma roth_number_nat_five : roth_number_nat 5 = 4 := rfl
+@[simp] lemma roth_number_nat_six : roth_number_nat 6 = 4 := rfl
+@[simp] lemma roth_number_nat_seven : roth_number_nat 7 = 4 := rfl
+@[simp] lemma roth_number_nat_eight : roth_number_nat 8 = 4 := rfl
+@[simp] lemma roth_number_nat_nine : roth_number_nat 9 = 5 := rfl
+@[simp] lemma roth_number_nat_ten : roth_number_nat 10 = 5 := dec_trivial
+@[simp] lemma roth_number_nat_eleven : roth_number_nat 11 = 6 :=
+begin
+  apply le_antisymm,
+  { simpa using roth_number_nat_add_le 3 8 },
+  apply add_salem_spencer.le_roth_number_nat {0, 1, 3, 7, 8, 10},
+  { dec_trivial },
+  { simp },
+  { simp }
+end
+
+-- `dec_trivial` times out on `roth_number_nat 12 = 6` with the current decidability instance
+
+@[simp] lemma roth_number_nat_thirteen : roth_number_nat 13 = 7 :=
+begin
+  apply le_antisymm,
+  { simpa using roth_number_nat_add_le 10 3 },
+  apply add_salem_spencer.le_roth_number_nat {0, 1, 3, 4, 9, 10, 12},
+  { dec_trivial },
+  { simp },
+  { simp }
+end
+
+@[simp] lemma roth_number_nat_fourteen : roth_number_nat 14 = 8 :=
+begin
+  apply le_antisymm,
+  { simpa using roth_number_nat_add_le 11 3 },
+  apply add_salem_spencer.le_roth_number_nat {0, 1, 3, 4, 9, 10, 12, 13}, -- unique example
+  { dec_trivial },
+  { simp },
+  { simp }
+end
+
+lemma roth_number_nat_dec_upper_bound {N M : ℕ} (h₂ : roth_number_nat N ≤ M)
```

Comment:
This was an attempt to get faster Roth number computations... Does it work to give an upper bound on roth 12? If not, it's probably not very helpful for anything else

Reply:
It works!